src/class/macros.rs

```diff
@@ -415,32 +389,51 @@ macro_rules! py_func_set_del {
 
             let py = $crate::Python::assume_gil_acquired();
             let _pool = $crate::GILPool::new(py);
-            let slf = py.mut_from_borrowed_ptr::<$generic>(slf);
+            let slf = py.from_borrowed_ptr::<$crate::PyCell<$generic>>(slf);
             let name = py.from_borrowed_ptr::<$crate::types::PyAny>(name);
 
             let result = if value.is_null() {
-                match name.extract() {
-                    Ok(name) => slf.$fn_del(name).into(),
-                    Err(e) => Err(e.into()),
-                }
+                call_refmut!(slf, $fn_del, name)
             } else {
                 let value = py.from_borrowed_ptr::<$crate::types::PyAny>(value);
-                match name.extract() {
-                    Ok(name) => match value.extract() {
-                        Ok(value) => slf.$fn_set(name, value).into(),
-                        Err(e) => Err(e.into()),
-                    },
-                    Err(e) => Err(e.into()),
-                }
+                call_refmut!(slf, $fn_set, name, value)
             };
             match result {
                 Ok(_) => 0,
-                Err(e) => {
-                    e.restore(py);
-                    -1
-                }
+                Err(e) => e.restore_and_minus1(py),
             }
         }
         Some(wrap::<$generic>)
     }};
 }
+
+// Utitlities for extract arguments + call method for PyCell<T>
+macro_rules! call_ref {
+    ($slf: expr, $fn: ident $(; $args: expr)*) => {
+        match $slf.try_borrow_unguarded() {
```

Comment:
I tried to change this to `try_borrow`, but it's difficult because of a lifetime problem.
In detail:
Now this macro is expanded to
```rust
match arg1.extract() {
    Ok(arg1) => match arg2.extract() => 
       Ok(arg2) => match slf.extract() {... }
       Err(e) => ...,
    Err(e) => ...,
}
```
and compiler claims that slf lives shorter than args.
So re-ordering this to
```rust
match slf.extract() {
    Ok(slf) => match arg1.extract() => { ... },
    Err(e) => ...,
}
```
works for most cases.., but still doesn't work well with `CallbackConverter`.